### ISSUE TITLE
Using CFFI to create bytes objects?

### ISSUE THREAD

**Issue by JamesTheAwesomeDude** [python-cffi/cffi#62]
When a large `bytes` object needs to be created based off C library functions, but really should not be mutated after creation, what's the best way to approach that?

The only work-arounds I'm *aware* of are

1. Unnecessarily copy the data with `bytes(ffi_owned_cdata)`
2. Just use `bytearray` instead of `ffi.new`

Ideally, I'd imagine an interface like this:

```python
from _libexample_cffi import ffi, lib

def foo() -> bytes:
  # 1. Allocate a freshly-made PyBytes object
  result_wip = ffi.new_bytes_builder(lib.MAX_LENGTH, resizable=True)
  # -- Internally calls PyBytes_FromStringAndSize(NULL, size)
  
  # 2. Use C library functions to build the object
  length = ffi.new('size_t*')
  lib.foo(result_wip, length)  # cdef('void foo(uint8_t* buf, size_t* len_written);')
  # -- cdata result_wip works as a buffer
  assert length[0] <= lib.MAX_LENGTH
  result_wip.resize(length[0])
  # -- Internally calls _PyBytes_Resize()
  
  # 3. Expose the PyBytes object that has been constructed
  result = bytesbuilder.finalize()
  # -- Does not call any Python API functions
  # -- Exposes result in-place, without a copy
  
  # - `result` is immutable bytes; further attempts to access cdata result_wip will raise a TypeError
  # - If result_wip is garbage-collected before finalization, it is deallocated, but this is not recommended
  # - If result_wip is used as a ContextManager and exited before finalization, it is deallocated
  
  assert type(result) is bytes
  return result
```

Does something like that already exist? Is it planned? Is it on the "wishlist", but low-priority / contributions-welcome?

**Comment by JamesTheAwesomeDude** [comment]
I tried adding a helper function that returned an already-filled-in `bytes` `PyObj`, but I had trouble exfiltrating the created `PyObject` *out* of my code:

```
result_cdata = lib.foo_PyBytes_wrapper()
result = ffi.from_handle(result_cdata)
# TypeError: expected a 'cdata' object with a 'void *' out of new_handle(), got 'PyObject *'
```

**Comment by arigo** [comment]
The problem of such an API is that it is very CPython-centric.  It is hard to reasonably get the no-copying effect on other Python implementations from that API.  Not necessarily impossible, but it's certainly something that we need to think about (even if only a subset of CFFI users care about PyPy at all, for example).

On a similar note, CFFI is not meant for manipulation of `PyObject *` or calling any of the CPython C API functions.  There are some ways to abuse it to do so, but it is in principle not supported at all, and may randomly crash for various reasons (for example, the GIL is not held when the CPython C API function is called via CFFI).

**Comment by JamesTheAwesomeDude** [comment]
I see. Thanks for clarifying that my current approach would be CPython-centric.

Is there any way to correctly fill data into a newly constructed Python `bytes` object using C code, in an implementation-agnostic way? Or is there no way out of this dilemma?:

> * Construct and return a `bytearray` object instead
> * Perform a wasteful copying cast to `bytes` upon return

**Comment by arigo** [comment]
No, for now you can't build a `bytes` object from cffi-invoked C code.  You need to have C code fill some other kind of object, and optionally copy it to a bytes afterwards.  It can be a `bytearray` or any kind of buffer like `array.array()` or `ffi.new()` etc.

I could hand-wave an explanation along the lines of "for small `bytes` then the extra copying is fast because it's all in caches" and "for very large amounts of data `bytes` is not really the most performant way to store data anyway".

The truth, though, is that it would be cool if there was a way, but there isn't so far.  Someone would need to look again inside PyPy for how a similar result could be achieved there, and from there we could check what kind of CFFI API would make sense for both PyPy and CPython.

**Comment by arigo** [comment]
Some object types that seem kind-of-equivalent on CPython may have different performance characteristics on PyPy: `bytes` are implemented completely differently than `array.array()` or `ffi.new("char[]", number_of_bytes)`.  The last two are examples that might be better in your case.

The issue is that it's hard to come up with a single design that maximizes performance for both very small and very large byte strings.  CPython uses a simple enough approach where the trade-off doesn't really show up; but PyPy is a lot faster with small objects.  For these, PyPy's moving garbage collector costs on the order of *a single CPU instruction* per object on average, including both allocation and later figuring out that the object is no longer used and freeing it.  The drawback to that is that we can't just "take a pointer" to the content of a `bytes`: the memory might move around under our feet.

In some cases, CFFI on PyPy will try to take a pointer anyway.  For example, if you pass a `bytes` to `ffi.myfunc()`, and the bytes is large enough, then it won't move (only small objects move) and we can take a pointer to it directly.  If that fails, then the `bytes` was small anyway, so the extra copy that CFFI needs to make is not too costly.  But that's only OK for read-only access to `bytes`, not read-write.

Anyway, all that was explanations for why some restrictions in CFFI may be unexpected if you're only used to CPython.  Now, it's probably possible to implement what you suggest in PyPy, but it's just not done so far.  I think that CFFI on PyPy could use `malloc()` to get an object, reserve some header space, and give the pointer to the user; and later some `bytesbuilder.finalize()` call adds the correct object headers for the GC and register it as a GC object, pretending it was a "large" object all along, to be disposed of by `free()`...
